### PR TITLE
✨ Add bastion host to login the nodes and update non HA template

### LINKED ISSUE #562: Login to the nodes using the bastion node like AWS
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Securely login to the nodes using the bastion node like AWS.
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/accessing-instances.md#accessing-nodes-via-ssh

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

This issue comes from #561 .

### PR DESCRIPTION
**Pull request by hidekazuna**
**What this PR does / why we need it**:

This PR adds the bastion host to login the nodes like [CAPA](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/master/docs/accessing-instances.md#enabling-the-bastion-host).

Update templates/cluster-template-without-lb.yaml to meet real use case.
-   managedSecurityGroups false to true
-   disablePortSecurity: true to false 

Update CI to use bastion.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #562 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note
Add the bastion host to login the nodes like CAPA.

```

### CODE REVIEW

**Comment by jichenjc** [comment]
just a quick question / discussion
is it possible to add this into CI ? e.g create a bastion into the cluster creation stage?

**Comment by hidekazuna** [comment]
> 
> 
> just a quick question / discussion
> is it possible to add this into CI ? e.g create a bastion into the cluster creation stage?

Yes, it is. We need:
- Update  hack/ci/e2e-conformance/e2e-conformance_patch.yaml.tpl for adding bastion to the manifest.
- Update hack/ci/e2e-conformance.sh to use bastion

**Comment by hidekazuna** [comment]
Changed title with WIP since I added CI update.

**Comment by hidekazuna** [comment]
@jichenjc CI updated to use bastion!

**Comment by jichenjc** [comment]
@hidekazuna thanks!
I will check it and reply soon

**Review by jichenjc — COMMENTED** [review]
some minor comments , nothing special~ thanks for the PR~

File: pkg/cloud/services/compute/bastion.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	infrav1 "sigs.k8s.io/cluster-api-provider-openstack/api/v1alpha3"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/record"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+)
+
+const (
+	DefaultSecurityGroup        = "default"
+	DefaultFailureDomain        = "nova"
+	DefaultSSHKeyName           = "default"
```
[the remark this answers is not in the document]
Comment (hidekazuna) [reply pkg/cloud/services/compute/bastion.go]: These are removed.

File: pkg/cloud/services/compute/bastion.go [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	// Get image ID
+	imageID, err := getImageID(s, openStackCluster.Spec.Bastion.Image)
+	if err != nil {
+		return nil, fmt.Errorf("create new server err: %v", err)
+	}
+	flavorName := openStackCluster.Spec.Bastion.Flavor
+	keyName := openStackCluster.Spec.Bastion.SSHKeyName
+	if keyName == "" {
+		keyName = DefaultSSHKeyName
+	}
```
[the remark this answers is not in the document]
Comment (hidekazuna) [reply pkg/cloud/services/compute/bastion.go]: Yes, you are right. Simply removed not given case as in instance.go.

File: pkg/cloud/services/compute/bastion.go [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+		record.Warnf(openStackCluster, "FailedCreateServer", "Failed to create bastion: %v", err)
+		return nil, fmt.Errorf("create new server err: %v", err)
+	}
+	record.Eventf(openStackCluster, "SuccessfulCreateServer", "Created server %s with id %s", name, server.ID)
+
+	return &Instance{Server: *server, State: infrav1.InstanceState(server.Status)}, nil
+
+}
+
+func getTimeout(name string, timeout int) time.Duration {
```
[the remark this answers is not in the document]
Comment (hidekazuna) [reply pkg/cloud/services/compute/bastion.go:139]: As mentioned above, we can not use instance.go now.

File: pkg/cloud/services/compute/bastion.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+		})
+		if err != nil {
+			return nil, errors.Errorf("error creating Openstack instance: %v", err)
+		}
+	}
+
+	return instance, nil
+}
+
+func (s *Service) DeleteBastion(clusterName string, openStackCluster *infrav1.OpenStackCluster) error {
```
[the remark this answers is not in the document]
Comment (hidekazuna) [reply pkg/cloud/services/compute/bastion.go:71]: bastion is manged in OpenStackCluster. Instances are managed by OpenStackMachine.
The code create/delete instance now includes OpenStackMachine.Spec. we can not use the code for bastion.

**Comment by hidekazuna** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
https://logs.openlabtesting.org/logs/19/619/1a927b8c51748f4863a975d68db7b577522f9de2/check/cluster-api-provider-openstack-current-acceptance-test-v1.18/b8d9e5c/job-output.txt.gz

```
020-09-02 04:17:04.385339 | ubuntu-bionic-large | Running virtualenv with interpreter /usr/bin/python3.6
2020-09-02 04:17:04.639379 | ubuntu-bionic-large | Traceback (most recent call last):
2020-09-02 04:17:04.639597 | ubuntu-bionic-large |   File "/usr/local/lib/python3.6/dist-packages/virtualenv.py", line 24, in <module>
2020-09-02 04:17:04.639661 | ubuntu-bionic-large |     import distutils.spawn
2020-09-02 04:17:04.639779 | ubuntu-bionic-large |   File "/usr/local/lib/python3.6/dist-packages/_distutils_hack/__init__.py", line 82, in create_module
2020-09-02 04:17:04.645680 | ubuntu-bionic-large |     return importlib.import_module('._distutils', 'setuptools')
2020-09-02 04:17:04.645803 | ubuntu-bionic-large |   File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
2020-09-02 04:17:04.647137 | ubuntu-bionic-large |     return _bootstrap._gcd_import(name[level:], package, level)
2020-09-02 04:17:04.647219 | ubuntu-bionic-large | ModuleNotFoundError: No module named 'setuptools'
2020-09-02 04:17:04.685334 | ubuntu-bionic-large | ./safe-devstack-vm-gate-wrap.sh: line 518: /tmp/ansible/bin/pip: No such file or directory
2020-09-02 04:17:04.686012 | ubuntu-bionic-large | ./safe-devstack-vm-gate-wrap.sh: line 519: /tmp/ansible/bin/pip: No such file or directory
2020-09-02 04:17:04.689069 | ubuntu-bionic-large | cat: /etc/nodepool/primary_node_private: No such file or directory
2020-09-02 04:17:04.690861 | ubuntu-bionic-large | cat: /etc/nodepool/sub_nodes_private: No such file or directory
2020-09-02 04:17:04.694733 | ubuntu-bionic-large | ./safe-devstack-vm-gate-wrap.sh: line 556: /tmp/ansible/bin/ansible: No such file or directory
2020-09-02 04:17:04.695697 | ubuntu-bionic-large | ./safe-devstack-vm-gate-wrap.sh: line 558: /tmp/ansible/bin/ansible: No such file or directory
2020-09-02 04:17:04.696579 | ubuntu-bionic-large | ./safe-devstack-vm-gate-wrap.sh: line 562: /tmp/ansible/bin/ansible: No such file or directory
2020-09-02 04:17:04.697423 | ubuntu-bionic-large | ./safe-devstack-vm-gate-wrap.sh: line 564: /tmp/ansible/bin/ansible: No such file or directory
2020-09-02 04:17:04.707647 | ubuntu-bionic-large | /home/zuul/workspace/devstack-gate/functions.sh: line 172: warning: command substitution: ignored null byte in input
2020-09-02 04:17:04.719614 | ubuntu-bionic-large | Setting up the hosts
2020-09-02 04:17:04.719789 | ubuntu-bionic-large | ... this takes a few seconds (logs at logs/devstack-gate-setup-host.txt.gz)
2020-09-02 04:17:04.721337 | ubuntu-bionic-large | ./safe-devstack-vm-gate-wrap.sh: line 578: /tmp/ansible/bin/ara: No such file or directory
2020-09-02 04:17:04.722594 | ubuntu-bionic-large | gzip: /home/zuul/workspace/logs/ara: No such file or directory
2020-09-02 04:17:04.722769 | ubuntu-bionic-large | *** FAILED with status: 127
2020-09-02 04:17:14.958014 | ubuntu-bionic-large | ERROR
2020-09-02 04:17:14.958502 | ubuntu-bionic-large | {
2020-09-02 04:17:14.958609 | ubuntu-bionic-large |   "delta": "0:00:05.828767",
2020-09-02 04:17:14.958701 | ubuntu-bionic-large |   "end": "2020-09-02 04:17:04.723643",
2020-09-02 04:17:14.958806 | ubuntu-bionic-large |   "msg": "non-zero return code",
2020-09-02 04:17:14.958906 | ubuntu-bionic-large |   "rc": 127,
2020-09-02 04:17:14.959017 | ubuntu-bionic-large |   "start": "2020-09-02 04:16:58.894876"
2020-09-02 04:17:14.959127 | ubuntu-bionic-large | }
```

seems related to openlab...

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by jichenjc** [comment]
```
2020-09-08 03:00:50.503148 | ubuntu-bionic-large | + local 'cmd=sudo journalctl --no-pager -u containerd.service'
2020-09-08 03:00:50.503449 | ubuntu-bionic-large | + ssh_key_pem=/tmp/capi-quickstart-key.pem
2020-09-08 03:00:50.504022 | ubuntu-bionic-large | + ssh_params='-o LogLevel=quiet -o ConnectTimeout=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no'
2020-09-08 03:00:50.507286 | ubuntu-bionic-large | + scp -o LogLevel=quiet -o ConnectTimeout=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i /tmp/capi-quickstart-key.pem /tmp/capi-quickstart-key.pem ubuntu@172.24.5.167:/tmp/capi-quickstart-key.pem
2020-09-08 03:01:03.289662 | ubuntu-bionic-large | scp: /tmp/capi-quickstart-key.pem: Permission denied
```
not sure this is the root cause. .. but seems test keep failing @hidekazuna

**Comment by hidekazuna** [comment]
> ```
> 2020-09-08 03:00:50.503148 | ubuntu-bionic-large | + local 'cmd=sudo journalctl --no-pager -u containerd.service'
> 2020-09-08 03:00:50.503449 | ubuntu-bionic-large | + ssh_key_pem=/tmp/capi-quickstart-key.pem
> 2020-09-08 03:00:50.504022 | ubuntu-bionic-large | + ssh_params='-o LogLevel=quiet -o ConnectTimeout=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no'
> 2020-09-08 03:00:50.507286 | ubuntu-bionic-large | + scp -o LogLevel=quiet -o ConnectTimeout=30 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -i /tmp/capi-quickstart-key.pem /tmp/capi-quickstart-key.pem ubuntu@172.24.5.167:/tmp/capi-quickstart-key.pem
> 2020-09-08 03:01:03.289662 | ubuntu-bionic-large | scp: /tmp/capi-quickstart-key.pem: Permission denied
> ```

These logs are shown in the successful test:
https://logs.openlabtesting.org/logs/17/617/8f7b92f7b277b913d1ef6f81d5e25251c781a4f8/check/cluster-api-provider-openstack-current-acceptance-test-v1.18/24a9121/job-output.txt.gz

Simply worker node did not become Running phase as always in the last failure test.
```
2020-09-08 02:25:43.057423 | ubuntu-bionic-large | NAME                                    PROVIDERID                                         PHASE
2020-09-08 02:25:43.083870 | ubuntu-bionic-large | capi-quickstart-control-plane-ljlhc     openstack://14d77ea1-1c41-431c-8162-7a2c1177b7d2   Running
2020-09-08 02:25:43.084012 | ubuntu-bionic-large | capi-quickstart-md-0-64fd6b5948-cmmd8   openstack://699a66c8-c0c4-4cb0-a107-2895cfed7b40   Provisioning
2020-09-08 02:25:43.620623 | ubuntu-bionic-large | cluster did not start in 30 mins ... bailing out!
2020-09-08 02:25:43.632409 | ubuntu-bionic-large | + [[ false == \t\r\u\e ]]
```

**Comment by hidekazuna** [comment]
recheck

**Comment by jichenjc** [comment]
you are right, I didn't check logs above

usually such bail out cost 90 min (see other checks),  and it take2 120 min now, that's why I thought new error occurs...

**Comment by hidekazuna** [comment]
recheck

**Comment by jichenjc** [comment]
recheck

**Comment by hidekazuna** [comment]
@jichenjc I added another commit bastion uses cirros image with tiny flavor. Because bastion node with m1.medium flavor may be stress for CI environment.

**Comment by jichenjc** [comment]
@hidekazuna good points ~, that explained why it's always fail while the other test has pass rate..

**Comment by hidekazuna** [comment]
recheck

**Pull request merged**